src/java/org/apache/cassandra/streaming/StreamManager.java

```diff
@@ -213,6 +226,98 @@ private static double calculateEffectiveRateInBytes(double throughput)
     private final Map<UUID, StreamResultFuture> followerStreams = new NonBlockingHashMap<>();
 
+    private final Map<UUID, StreamingState> states = new NonBlockingHashMap<>();
+    private volatile ScheduledFuture<?> cleanup = null;
+
+    public StreamManager()
+    {
+        addListener(new StreamListener()
+        {
+            @Override
+            public void onRegister(StreamResultFuture result)
+            {
+                StreamingState state = new StreamingState(result);
+                StreamingState previous = states.putIfAbsent(state.getId(), state);
+                if (previous == null)
+                {
+                    state.phase.start();
+                    result.addEventListener(state);
+                }
+                else
+                {
+                    logger.warn("Duplicate streaming states detected for id {}", state.getId());
+                }
+            }
+        });
+    }
+
+    public void start()
+    {
+        this.cleanup = ScheduledExecutors.optionalTasks.scheduleAtFixedRate(this::cleanup, 0,
+                                                             DatabaseDescriptor.getStreamingStateCleanupInterval().toNanoseconds(),
+                                                             TimeUnit.NANOSECONDS);
+    }
+
+    public void stop()
+    {
+        ScheduledFuture<?> cleanup = this.cleanup;
+        if (cleanup != null)
+            cleanup.cancel(false);
+    }
+
+    public Collection<StreamingState> getStreamingStates()
+    {
+        return states.values();
+    }
+
+    public StreamingState getStreamingState(UUID id)
+    {
+        return states.get(id);
+    }
+
+    @VisibleForTesting
+    public void putStreamingState(StreamingState state)
+    {
+        StreamingState previous = states.putIfAbsent(state.getId(), state);
+        if (previous != null)
+            throw new AssertionError("StreamPlan id " + state.getId() + " already exists");
+    }
+
+    @VisibleForTesting
+    public void cleanup()
+    {
+        try
+        {
+            DurationSpec duration = DatabaseDescriptor.getStreamingStateExpires();
+            long durationNanos = duration.toNanoseconds();
+            long deadlineNanos = Clock.Global.nanoTime() - durationNanos;
+            for (StreamingState state : states.values())
+            {
+                if (state.getLastUpdatedAtNanos() < deadlineNanos)
+                {
+                    if (state.isComplete())
+                    {
+                        states.remove(state.getId());
+                    }
+                    else
+                    {
+                        logger.warn("Stream {} has been running longer than state expires window {};\n{}", state.getId(), duration, state);
+                    }
+                }
+            }
+        }
+        catch (Throwable t)
+        {
+            logger.warn("Unexpected error cleaning up stream state", t);
```

Comment:
with scheduled tasks, if the task throws the scheduling stops, so using this to make sure we un again even if there is an error